internal/services/conditionalaccess/named_location_resource.go

```diff
@@ -0,0 +1,423 @@
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/manicminer/hamilton/msgraph"
+	"github.com/manicminer/hamilton/odata"
+
+	"github.com/hashicorp/terraform-provider-azuread/internal/clients"
+	"github.com/hashicorp/terraform-provider-azuread/internal/tf"
+	"github.com/hashicorp/terraform-provider-azuread/internal/utils"
+	"github.com/hashicorp/terraform-provider-azuread/internal/validate"
+)
+
+func namedLocationResource() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: namedLocationResourceCreate,
+		ReadContext:   namedLocationResourceRead,
+		UpdateContext: namedLocationResourceUpdate,
+		DeleteContext: namedLocationResourceDelete,
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(5 * time.Minute),
+			Read:   schema.DefaultTimeout(5 * time.Minute),
+			Update: schema.DefaultTimeout(5 * time.Minute),
+			Delete: schema.DefaultTimeout(5 * time.Minute),
+		},
+
+		Importer: tf.ValidateResourceIDPriorToImport(func(id string) error {
+			if _, err := uuid.ParseUUID(id); err != nil {
+				return fmt.Errorf("specified ID (%q) is not valid: %s", id, err)
+			}
+			return nil
+		}),
+
+		Schema: map[string]*schema.Schema{
+
+			"display_name": {
+				Type:             schema.TypeString,
+				Required:         true,
+				ValidateDiagFunc: validate.NoEmptyStrings,
+			},
+
+			"ip": {
+				Type:         schema.TypeList,
+				Optional:     true,
+				ForceNew:     true,
+				MaxItems:     1,
+				ExactlyOneOf: []string{"ip", "country"},
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"ip_ranges": {
+							Type:     schema.TypeList,
+							Required: true,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
+						},
+
+						"trusted": {
+							Type:     schema.TypeBool,
+							Optional: true,
+						},
+					},
+				},
+			},
+
+			"country": {
+				Type:         schema.TypeList,
+				Optional:     true,
+				ForceNew:     true,
+				MaxItems:     1,
+				ExactlyOneOf: []string{"ip", "country"},
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"countries_and_regions": {
+							Type:     schema.TypeList,
+							Required: true,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
+						},
+
+						"include_unknown_countries_and_regions": {
+							Type:     schema.TypeBool,
+							Optional: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func namedLocationResourceCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*clients.Client).ConditionalAccess.NamedLocationsClient
+
+	displayName := d.Get("display_name").(string)
+
+	base := msgraph.BaseNamedLocation{
+		DisplayName: utils.String(displayName),
+	}
+
+	if v, ok := d.GetOk("ip"); ok {
+		properties := expandIPNamedLocation(v.([]interface{}))
+		properties.BaseNamedLocation = &base
+
+		location, _, err := client.CreateIP(ctx, *properties)
```

Comment:
I think it would be easier to understand this if `location` was renamed to `ipLocation` here and `countryLocation` in line 133.